Auth tutorials (#1228)
streamlit/docs: content/develop/tutorials/authentication/microsoft.md

---
title: Use Microsoft Entra to authenticate users
slug: /develop/tutorials/authentication/microsoft
description: Learn how to authenticate users with Microsoft Entra
---

# Use Microsoft Entra to authenticate users

> [Microsoft Identity Platform](https://learn.microsoft.com/en-us/entra/identity-platform/v2-overview) is a service within Microsoft Entra that lets you build applications to authenticate users. Your applications can use personal, work, and school accounts managed by Microsoft. You can also connect other identity providers, like [Facebook](https://learn.microsoft.com/en-us/entra/external-id/facebook-federation).

Reviewer: > You can also connect with other identity providers, like [Facebook]
I am not sure it will work out of the box with the OIDC flow described in the tutorial, where the actual Auth flow will happen, I assume on Azure, and it will talk with Facebook in the background?

Please ignore this comment, if it has already been checked :)

Author: I'll remove it to be safe.

1. Save your `secrets.toml` file.

### Initialize your app

1. In `your_repository`, create a file named `app.py`.
1. In a terminal, change directories to `your_repository`, and start your app:

   ```bash
   streamlit run app.py
   ```

   Your app will be blank because you still need to add code.

1. In `app.py`, write the following:

   ```python
   import streamlit as st
   ```

1. Save your `app.py` file, and view your running app.
1. In your app, select "**Always rerun**", or press the "**A**" key.

   Your preview will be blank but will automatically update as you save changes to `app.py`.

1. Return to your code.

### Log the user in and out

1. Define a function that prompts the user to log in:

   ```python
   def login_screen():
       st.header("This app is private.")
       st.subheader("Please log in.")
       st.button("Log in with Microsoft", on_click=st.login)
   ```

   This function displays a short message and a button. Streamlit's login command is assigned to the button as a callback. Alternatively, you could use a conditional:

   ```python
       if st.button("Log in with Google"):
           st.login()
   ```

> 1. If no user is logged in, call your function to prompt the user:

Reviewer 2: See my suggestion in the comment under line 222 of the authentication/google file.

   ```python
   if not st.experimental_user.is_logged_in:
       login_screen()
   ```

1. Otherwise, if a user _is_ logged in, display their information:

   ```python
   else:
       st.experimental_user
   ```

   Since `st.experimental_user` is a dict-like object in a line by itself, Streamlit magic displays it in your app.

1. Save your `app.py` file, and test your running app.

>    In your live preview, if you log in to your app, the login button will be replaced with the contents of your identity token. Observe the different values that are available from Google.

Reviewer 2: ```suggestion
   In your live preview, when you log in to your app, the login button is replaced with the contents of your identity token. Observe the different values that are available from Google.
```
What's the reason for observing the different values? Just curiosity, or should the user learn something from this?

1. Return to your code.

1. Replace `st.experimental_user` with a personalized greeting:

   ```diff
   else:
   -   st.experimental_user
   +   st.header(f"Welcome, {st.experimental_user.name}!")
   ```

1. Add a logout button:

   ```python
       st.button("Log out", on_click=st.logout)
   ```

1. Save your `app.py` file and test your running app.

   In your live preview, if you log out of your app, it will return to the login prompt.

## Deploy your app on Community Cloud

When you are ready to deploy your app, you must update your application in Microsoft Azure and your secrets. The following steps describe how to deploy your app on Community Cloud.

1. Add a `requirements.txt` file to your repository with the following lines:

   ```txt
   streamlit>=1.42.0
   Authlib>=1.3.2
   ```

>    This ensures the correct Python dependencies are installed for your deployed app.

Reviewer 2: ```suggestion
   This ensures that the correct Python dependencies are installed for your deployed app.
```

1. Save your `requirements.txt` file.

> 1. Deploy your app and take note of your app's URL.

Reviewer 2: ```suggestion
1. Deploy your app, and copy your app's URL into a text editor.
```

   You'll update your secrets in the following steps. For more information about deploying an app on Community Cloud, see [Deploy your app](/deploy/streamlit-community-cloud/deploy-your-app).

1. In your [app settings](/deploy/streamlit-community-cloud/manage-your-app/app-settings) in Community Cloud, select "**Secrets**."

> 1. Copy the contents of your local `secrets.toml` file and paste them into your app settings.

Reviewer 2: ```suggestion
1. Copy the contents of your local `secrets.toml` file, and paste them into your app settings.
```
